### ISSUE TITLE
hydra: mpiexec jobs no longer run in background

### ISSUE THREAD

**Issue by mpichbot** [pmodels/mpich#2022]
**Originally by _huiweilu_ on 2014-02-10 14:28:21 -0600**

---

A shell script containing multiple mpiexec commands no longer makes it past the first one.  Apparently one now needs to set stdin to /dev/null to continue.

```
$ cat t.sh
#!/bin/sh

$MPIEXEC -n 1 hostname
$MPIEXEC -n 1 date

## MPICH-3.1b1 (also 3.0 series), "date" output not printed
$ MPIEXEC=/opt/mpich/bin/mpiexec ./t.sh &
[5] 29761
$ batura


[5]+  Stopped                 MPIEXEC=/opt/mpich/bin/mpiexec ./t.sh


## Open MPI - note that "date" output is printed
$ MPIEXEC=/usr/bin/mpiexec ./t.sh &
[6] 29780
$ batura
Sun Feb  9 14:12:35 MST 2014

[6]   Done                    MPIEXEC=/usr/bin/mpiexec ./t.sh
```

Reported by Jed Brown from https://lists.mpich.org/mailman/listinfo/discuss

**Comment by hzhou** [comment]
Just tested it with the current `main` branch, and it seems the issue no longer exist. @jedbrown Would you like to confirm before we close this issue?

**Comment by jedbrown** [comment]
I still see it with mpich-4.0. Please use my script as written. Pavan's comment above changed to have the script launch one background job and one foreground. My script is meant to be launched in the background, with both lines blocking inside the script.
```sh
#!/bin/sh

$MPIEXEC -n 1 uname
$MPIEXEC -n 1 date
```
Here, I press enter when it's hung after printing `Linux`.
```console
sh-5.1$ MPIEXEC=/opt/mpich/bin/mpiexec ./t.sh &
[1] 355649
sh-5.1$ Linux


[1]+  Stopped(SIGTTIN)        MPIEXEC=/opt/mpich/bin/mpiexec ./t.sh
sh-5.1$ jobs
[1]+  Stopped(SIGTTIN)        MPIEXEC=/opt/mpich/bin/mpiexec ./t.sh
```
This works with Open MPI:
```console
sh-5.1$ MPIEXEC=/usr/bin/mpiexec ./t.sh &
[1] 355710
sh-5.1$ Linux
Mon Feb 14 10:58:39 AM MST 2022
jobs
[1]+  Done                    MPIEXEC=/usr/bin/mpiexec ./t.sh
```

**Comment by hzhou** [comment]
I see. I managed to reproduce it on the Centos system. I was running it on my local ubuntu, which seems to work fine. Okay, I'll investigate.

**Comment by hzhou** [comment]
Hmm, I tried it again today and can't reproduce it --
```
[pmrs-gpu-240-02 ~/temp/mpich-5860] $ cat t.sh
#!/bin/sh
mpirun -n 1 uname
mpirun -n 1 date
[pmrs-gpu-240-02 ~/temp/mpich-5860] $ ./t.sh &
[1] 141785
[pmrs-gpu-240-02 ~/temp/mpich-5860] $ Linux
Tue Aug  9 14:28:51 CDT 2022

[1]+  Done                    ./t.sh
[pmrs-gpu-240-02 ~/temp/mpich-5860] $
```

**Comment by jedbrown** [comment]
Still a problem for me.
```console
$ MPIEXEC=/opt/mpich/bin/mpiexec ./t.sh  &
[1] 297763
$ Linux


[1]+  Stopped                 MPIEXEC=/opt/mpich/bin/mpiexec ./t.sh
$ j
[1]+ 297763 Stopped (tty input)     MPIEXEC=/opt/mpich/bin/mpiexec ./t.sh
```

This still works correctly with ompi.
```console
$ /opt/mpich/bin/mpichversion
MPICH Version:          4.0.2
MPICH Release date:     Thu Apr  7 12:34:45 CDT 2022
MPICH Device:           ch4:ucx
MPICH configure:        --prefix=/opt/mpich --with-device=ch4:ucx --with-hwloc --without-java --enable-error-checking=runtime --enable-error-messages=all --enable-g=meminit CC=gcc CXX=g++ FC=gfortran FFLAGS=-fallow-argument-mismatch FCFLAGS=-fallow-argument-mismatch
MPICH CC:       gcc  -march=native -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -Wno-error=array-bounds  -O2
MPICH CXX:      g++  -march=native -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -Wp,-D_GLIBCXX_ASSERTIONS -O2
MPICH F77:      gfortran -fallow-argument-mismatch  -O2
MPICH FC:       gfortran -fallow-argument-mismatch  -O2
MPICH Custom Information:
```

**Comment by hzhou** [comment]
@jedbrown I am checking the current main branch. Somewhere between `v4.0` and current `main` fixes it.

**Comment by jedbrown** [comment]
Okay, I can confirm that 4.1a1 works for me.

**Comment by hzhou** [comment]
Bisecting a fix is weird -- "good" means "bad" and "bad" means "good" 😄 
```
e4ccac1679217724ea8f59a48b20b0422a079db5 is the first bad commit
commit e4ccac1679217724ea8f59a48b20b0422a079db5
Author: Ken Raffenetti <raffenet@mcs.anl.gov>
Date:   Thu Jan 6 11:09:36 2022 -0600

    pm/hydra: Remove unnecessary signal handler

    All this signal handler did was set a global variable that was never
    read again.
```
Definitely an unsuspecting fix.

**Comment by hzhou** [comment]
The commit https://github.com/pmodels/mpich/pull/5751/commits/e4ccac1679217724ea8f59a48b20b0422a079db5
And the man page on `SIGTTIN`:
```
Macro: int SIGTTIN

    A process cannot read from the user’s terminal while it is running as a background job. When any process in a background job tries to read from the terminal, all of the processes in the job are sent a SIGTTIN signal. The default action for this signal is to stop the process. For more information about how this interacts with the terminal driver, see [Access to the Controlling Terminal](https://www.gnu.org/software/libc/manual/html_node/Access-to-the-Terminal.html).
```

**Comment by jedbrown** [comment]
Hah, that explains it. Thanks for investigating, and thanks @raffenet for the inadvertent fix. Seems we can close this issue now; fixed in v4.1 (which I assume is coming soonish).

**Comment by hzhou** [comment]
It is weird that in this case, we want `SIGTTIN` to kill the process. I guess sending SIGTTIN is the mechanism how shell works in this case.

**Comment by raffenet** [comment]
We are still ignoring `SIGTTIN` after the patch, just not with the custom handler. I'm surprised it made any functional difference, but maybe I'm not seeing the whole picture. I should clean up the comments in that function, because they don't really match the code any more.

**Comment by hzhou** [comment]
I am not entirely sure about it either. I'll try to dig it a bit deeper...

**Comment by raffenet** [comment]
It is possible that the `read` call was just restarting endlessly?
```
For some system calls, if a signal is caught while the call is executing and the call is prematurely terminated, the
     call is automatically restarted.  Any handler installed with signal(3) will have the SA_RESTART flag set, meaning that
     any restartable system call will not return on receipt of a signal.  The affected system calls include read(2),
     write(2), sendto(2), recvfrom(2), sendmsg(2), and recvmsg(2) on a communications channel or a low speed device and
     during a ioctl(2) or wait(2).  However, calls that have already committed are not restarted, but instead return a
     partial success (for example, a short read count).  These semantics could be changed with siginterrupt(3).
```

**Comment by raffenet** [comment]
FWIW, the zero byte `read` call does not seem to reliably do what the comment says it should, which is return an error with `errno = EINTR` when our handler was installed. On my macOS laptop, there is no SIGTTIN when doing zero byte `read` from `stdin` when in the background. So the output of `HYDT_dmxi_stdin_valid` is questionable still...

**Comment by hzhou** [comment]
Here is what I gathered -
* The process stack is like:  `sh` (foreground)  --> `t.sh &` (background) --> `mpiexec` (background)
* When mpiexec test STDIN with read, `SIGTTIN` is generated
* `mpiexec` handles `SIGTTIN` and continue without issue and in fact exit normally
* _It appears_ `SIGTTIN` is delivered to both `t.sh &` and `mpiexec`. The latter handles it and continue to exit, but the former was put to `STOP`, which is the default action of `SIGTTIN`
* Setting `SIGTTIN` to `SIG_IGNORE` prevents the signal to be generated at all

Frankly, this appears to be a linux OS bug to me.

**Comment by raffenet** [comment]
> FWIW, the zero byte `read` call does not seem to reliably do what the comment says it should, which is return an error with `errno = EINTR` when our handler was installed. On my macOS laptop, there is no SIGTTIN when doing zero byte `read` from `stdin` when in the background. So the output of `HYDT_dmxi_stdin_valid` is questionable still...

I checked a toy program running in the background on macOS and Linux. On macOS, the zero byte read from stdin always succeeds regardless of signal handler settings. On Linux, if we use our custom handler, we get `errno=EINTR`. With the current `SIG_IGN` setting, we get back `errno=EIO`. The output of the function only cares if the call succeeds or fails and doesn't verify the errno setting, so I guess it still works as expected. I guess I got lucky because I really didn't think too hard when I was making the change 😅.
